### PR TITLE
enhance: initial solution for namespaced xml tags

### LINKED ISSUE #40: Fix namespaced xml tags
Original issue: https://github.com/scrapy/scrapy/issues/4833

### PR DESCRIPTION
**Pull request by kazaaz**
Co-authored-by: alkazaz alkazaz@kth.se
Co-authored-by: swill swill@kth.se
Co-authored-by: lerjevik lerjevik@kth.se
Co-authored-by: djiar djiar@kth.se
Co-authored-by: aljica aljica@kth.se

### CODE REVIEW

File: tests/test_utils_iterators.py [hunk 1]
```diff
@@ -232,10 +232,6 @@ def test_xmliter_encoding(self):
 class LxmlXmliterTestCase(XmliterTestCase):
     xmliter = staticmethod(xmliter_lxml)
 
-    @mark.xfail(reason='known bug of the current implementation')
-    def test_xmliter_namespaced_nodename(self):
-        super().test_xmliter_namespaced_nodename()
-
```
Comment (djikoSal) [inline tests/test_utils_iterators.py]: This should not be deleted fully, only the first line with @mark.xmail

**Review by djikoSal — CHANGES_REQUESTED** [review]
Not entirely correct

File: scrapy/utils/iterators.py [hunk 2]
```diff
@@ -53,12 +53,27 @@ def xmliter(obj, nodename):
         yield Selector(text=nodetext, type='xml')
 
 
+def _replace_all(obj, old, new):
+    if isinstance(obj, Response):
+        return obj.text.replace(old, new)
```
Comment (djikoSal) [inline scrapy/utils/iterators.py]:
```suggestion
        return obj.replace(body=obj.text.replace(old, new))
```
```suggestion
        return obj.text.replace(old, new)
```

File: scrapy/utils/iterators.py [hunk 3]
```diff
@@ -53,12 +53,27 @@ def xmliter(obj, nodename):
...
+    elif isinstance(obj, str):
+        return obj.replace(old, new)
+    elif isinstance(obj, bytes):
+        return str(obj).replace(old, new)
```
Comment (djikoSal) [inline scrapy/utils/iterators.py]:
```suggestion
        return obj.decode().replace(old, new).encode()
```
```suggestion
        return str(obj).replace(old, new)
```

File: scrapy/utils/iterators.py [hunk 4]
```diff
@@ -53,12 +53,27 @@ def xmliter(obj, nodename):
...
+
+
 def xmliter_lxml(obj, nodename, namespace=None, prefix='x'):
     from lxml import etree
-    reader = _StreamReader(obj)
-    tag = f'{{{namespace}}}{nodename}' if namespace else nodename
+    nodename_sanitized = nodename
+    obj_sanitized = obj
+    if nodename.find(":") > 0:
+        nodename_sanitized =  nodename.replace(":", "-")
```
Comment (djikoSal) [inline scrapy/utils/iterators.py]: Perhaps re-use our _replace_all here? e.g. do nodename._replace_all

**Review by djikoSal — CHANGES_REQUESTED** [review]

File: tests/test_utils_iterators.py (hunk 1 above)
Comment (kazaaz) [reply tests/test_utils_iterators.py]: Should be fixed now!

**Review by djikoSal — COMMENTED** [review]
Looks good now!

**Review by djikoSal — APPROVED** [review]

**Pull request closed without merging**